### ISSUE TITLE
TF2.0 - why does this keras model learn data which is explicitly avoided in the loss?

### ISSUE THREAD

**Issue by mauricio-fernandez-l** [tensorflow/tensorflow#36949]
I do not understand, how does a trained model learn values not influencing the loss function as given below by a minimal example. I am avoiding the error computation of corrupted data but my model still learns that corrupted data. I want to understand what is happening.

Suppose you have the following data for a function f(x) with input x (4 data points, input is 3-dimensional, output is 2-dimensional)

```
#%% Import
import numpy as np
import tensorflow as tf

#%% Data
x = np.array([
     [1,1,1]
     ,[1,1,2]
     ,[1,2,1]
     ,[1,2,2]
     ])

x = (x-np.mean(x,axis=0))

f = np.array([
     [5,10]
     ,[6,50]
     ,[4,-30]
     ,[7,-30]
     ])

d_f = 2
```

Suppose you want to train a keras model on this data, and the data values for f are corrupt in the last two data points in the second dimension. That means, I do NOT trust the number -30 (corrupted data). I still want to train a keras model. For that, I simply build a custom loss function which takes the mean squared error in the first 2 data points ([5,10] and [6,50]) and then only the square error in the last 2 data points ONLY IN THE FIRST COLUMN (from [4,-30] only [4] and from [7,-30] only [7]).

```
#%% Model with custom loss

model = tf.keras.Sequential([
    tf.keras.layers.Dense(16,activation='softplus')
    ,tf.keras.layers.Dense(16,activation='softplus')
    ,tf.keras.layers.Dense(d_f)
    ])

def myloss(f,f_m):
    f = tf.keras.backend.cast(f,tf.float32)
    ds = (f-f_m)**2
    ok = tf.keras.backend.mean(ds[0:2],axis=-1)
    corrupted = ds[2:,0]
    return tf.keras.backend.concatenate([ok,corrupted])

model.compile(
    optimizer='adam'
    ,loss=myloss
    )

print('Keras - evaluated loss:\n\t%.8f' % model.evaluate(x,f,verbose=0))
f_m = model.predict(x)
ds = (f-f_m)**2
myloss_output = np.concatenate([np.mean(ds[:2],axis=-1),ds[2:,0]])
print('Check f, f_m and myloss_output')
print(f)
print(f_m)
print(myloss_output)
print('Manually computed mean:\n\t%.8f' % np.mean(myloss_output))
```

My custom function myloss does what it should (mse w.r.t the first two data points and squared error w.r.t. the last two). The evaluation returns the mean over the computations, see code above. Deviations w.r.t. to the number -30 DO NOT APPEAR IN THE OUTPUT OF MYLOSS. The thing I do NOT understand here is that when I train the model, the trained model also approximates -30, although this is not in the output of myloss (see final code below).

**1) Why is that?**

**2) How can I prevent it?**

**3) How can I train the model with the provided data such that after training I can see, what the model learned by itself for the positions of the corrupted data (-30)?**

```
#%% Train and check predictions

model.fit(x,f,epochs=1500,verbose=2,batch_size=len(x))

# I do not understand, why the components f[2:,1] are ALSO trained, although the custom loss does not contain them
print(f)
print(model.predict(x))


# Output in console
[[  5  10]
 [  6  50]
 [  4 -30]
 [  7 -30]]
[[  4.269913   16.358786 ]
 [  6.6172767  45.615295 ]
 [  4.704413  -35.412888 ]
 [  6.408976  -25.318275 ]]
```

**Comment by mauricio-fernandez-l** [comment]
Sorry, got it, I just forgot about the default shuffle, i.e., 

`model.fit(x,f,epochs=1500,verbose=2,batch_size=len(x),shuffle=False)`

solves the problem! I will close the issue.
